get rid of TUn and TBin AST nodes completely

byorgey opened the pull request:
Unary and binary operations were getting desugared into applications
of operator primitives, but still initially parsed as separate special
TUn and TBin nodes.  This patch gets rid of the special TUn and TBin
nodes and now parses unary and binary operators directly as a
TApp (TPrim ...) ...  TUn and TBin are now pattern synonyms, so a lot
of code that used them does not have to change.

This is a preliminary step towards #185 and #136.

byorgey commented:
@shaylew can you see if this makes sense to you?

shaylew commented on src/Disco/Parser.hs:
```diff
@@ -877,12 +877,6 @@ parseExpr = (fixJuxtMul . fixChains) <$> (makeExprParser parseAtom table <?> "ex
     fixPrec (TBin bop1 t1 (TBin bop2 t2 t3))
       | bPrec bop2 < bPrec bop1 = TBin bop2 (fixPrec $ TBin bop1 t1 t2) t3
 
-    fixPrec (TApp (TBin bop t1 t2) t3)
-      | bPrec bop < funPrec = TBin bop t1 (fixPrec $ TApp t2 t3)
-
-    fixPrec (TApp t1 (TBin bop t2 t3))
-      | bPrec bop < funPrec = TBin bop (fixPrec $ TApp t1 t2) t3
-
```
Is the idea here that nothing will bind tighter than functions (in which case these are unreachable)? This makes sense, but it may also mean there's no reason for a configurable `funPrec` in the first place -- after this the parser won't use it, though the pretty printer still does.

byorgey replied:
I *think* this code was actually unreachable before, for exactly the reason you mention, but I had never noticed.  However, after making `TUn` and `TBin` into pattern synonyms, the second one became reachable!  In particular the second case matches when a lower-precedence unary operator is applied to a higher-precedence binary operator, and there's exactly one case where that can sensibly happen, when an exponentiation is negated.  In fact the reason I noticed is that precisely one expression in one single test started failing: -3^2 was evaluating as 9 instead of -9.   It took me a long time to track it down!

You may be right about not needing a configurable `funPrec`, I'll have to think about it more.

byorgey replied:
Actually I think `funPrec` is fine.  It is definitely needed in the pretty-printer. And it's not so much that it's configurable, as that it's automatically assigned to be one level higher than the highest operator.  If we ever add new operators with a precedence level in between existing ones, the `funPrec` will be automatically updated appropriately.

byorgey commented:
Oh, very good catch re: the dead code in the pretty-printer.  In fact, you caught a bug: you're right that the two code paths for unary operators (depending on whether it was a `TApp TPrim` or a `TUn`) were different, and in fact it turns out that factorials (our only postfix unary operator at the moment) were being pretty-printed incorrectly, e.g. `:pretty 5!` would show `!5`.  But the code for `TApp TPrim` already handled this correctly. So I added a test and deleted the dead code.

byorgey commented:
Thanks!
